streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/KafkaStreamsStateImpl.java

```diff
@@ -0,0 +1,140 @@
+
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Collections.unmodifiableSortedMap;
+import static java.util.Collections.unmodifiableSortedSet;
+import static java.util.Comparator.comparingLong;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.processor.assignment.KafkaStreamsState;
+import org.apache.kafka.streams.processor.assignment.ProcessId;
+import org.apache.kafka.streams.state.HostInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KafkaStreamsStateImpl implements KafkaStreamsState {
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaStreamsStateImpl.class);
+
+    private final ProcessId processId;
+    private final int numProcessingThreads;
+    private final Map<String, String> clientTags;
+    private final Map<TaskId, Long> taskLagTotals; // contains lag for all stateful tasks in the app topology
+    private final SortedSet<TaskId> previousActiveTasks;
+    private final SortedSet<TaskId> previousStandbyTasks;
+    private final SortedMap<String, Set<TaskId>> taskIdsByConsumer;
+    private final Optional<HostInfo> hostInfo;
+
+    public KafkaStreamsStateImpl(
+        final ProcessId processId,
+        final int numProcessingThreads,
+        final Map<String, String> clientTags,
+        final Map<TaskId, Long> taskLagTotals,
+        final SortedSet<TaskId> previousActiveTasks,
+        final SortedSet<TaskId> previousStandbyTasks,
+        final SortedMap<String, Set<TaskId>> taskIdsByConsumer,
+        final Optional<HostInfo> hostInfo
+    ) {
+        this.processId = processId;
+        this.numProcessingThreads = numProcessingThreads;
+        this.clientTags = unmodifiableMap(clientTags);
+        this.taskLagTotals = unmodifiableMap(taskLagTotals);
+        this.previousActiveTasks = unmodifiableSortedSet(previousActiveTasks);
+        this.previousStandbyTasks = unmodifiableSortedSet(previousStandbyTasks);
+        this.taskIdsByConsumer = unmodifiableSortedMap(taskIdsByConsumer);
+        this.hostInfo = hostInfo;
+    }
+
+    @Override
+    public ProcessId processId() {
+        return processId;
+    }
+
+    @Override
+    public int numProcessingThreads() {
+        return numProcessingThreads;
+    }
+
+    @Override
+    public SortedSet<String> consumerClientIds() {
+        return new TreeSet<>(taskIdsByConsumer.keySet());
+    }
+
+    @Override
+    public SortedSet<TaskId> previousActiveTasks() {
+        return previousActiveTasks;
+    }
+
+    @Override
+    public SortedSet<TaskId> previousStandbyTasks() {
+        return previousStandbyTasks;
+    }
+
+    @Override
+    public long lagFor(final TaskId task) {
+        final Long totalLag = taskLagTotals.get(task);
```

Comment:
We should also throw if the KafkaStreamsState was built without requesting the task lags be computed, right? (IIRC we decided on UnsupportedOperationException in the last PR)

Same for the other lag-related methods here. I guess the safest thing to do is wrap the `taskLagTotals` field in an Optional and make it empty when the user passed in `computeTaskLags=false` to the `ApplicationState#kafkaStreamsStates` API?